Automattic/woocommerce-subscriptions-core
# Fix issues causing caches on HPOS sites with sync off to not update

Fixes #283
Fixes #194

WC Subscriptions has a user subscriptions cache that is stored in user meta. An example of what this cache looks like is in the screenshot below. 

<img width="421" alt="Screen Shot 2022-11-25 at 9 30 03 am" src="https://user-images.githubusercontent.com/8490476/203874322-ad03e18e-20f2-4675-b759-82cf862018f3.png">

^ This customer (user ID 1) has 1 subscription with the ID 875. 

This is a persistent cache and not a cache that expires after x time has passed. Because it is persistent, we need to make sure it is updated whenever a subscription is created for a user, or the customer ID of a subscription changes. 

In a WP Posts world, this was done with a series of post update related hooks. eg `added_post_meta`,`update_post_meta`, `deleted_post_meta` etc. When those hooks fire, our `WCS_Post_Meta_Cache_Manager` class checks if it an update, add or delete for a meta-key we're interested in (`_customer_user` or `_subscription_renewal etc), and if it is, it fires an action hook that we listen too and update our caches appropriately. 

These hooks won't fire in HPOS environments with syncing turned off and so our persistent subscription cache falls down. 

This PR adds a new cache manager class to handle HPOS environments and instead of listening on post related hooks, it uses WC CRUD APIs.  

## Description

This PR lays the groundwork for introducing a `WCS_Object_Data_Cache_Manager` which extends the `WCS_Post_Meta_Cache_Manager` class.

Rather than listen on post related hooks, the new `WCS_Object_Data_Cache_Manager` listens on `woocommerce_before_{$this->object_type}_object_save` and `woocommerce_after_{$this->object_type}_object_save` CRUD hooks. 

When an object type (`'order` or `'subscription'`) is saved, this class will see if any meta, or property that it is interested in has changed, and if it has changed, will fire the same hook I mentioned above, and that hook will result in our caches being updated. 

This PR only lays the groundwork, further changes will be needed to handle when an order or subscription is trashed, deleted or untrashed, similar to how the `WCS_Post_Meta_Cache_Manager` works. There aren't equivalent hooks or ways to test this yet because of the following issues: 

- https://github.com/Automattic/woocommerce-subscriptions-core/issues/221
- https://github.com/Automattic/woocommerce-subscriptions-core/issues/195

Once we resolve the subscription trashed/deleted hooks, the new class (`WCS_Object_Data_Cache_Manager`) should add support for them. See issue below.

- https://github.com/Automattic/woocommerce-subscriptions-core/issues/310

## How to test this PR

1. With HPOS enabled and syncing turned off (this change changes nothing on WP Post environments). 
2. Purchase a subscription.
3. Go to the **My account → Subscriptions** page (/my-account/subscriptions/) 
4. Purchase a subscription.
3. Go to the **My account → Subscriptions** page (/my-account/subscriptions/)
    - On `trunk` you won't have both subscriptions listed, only the first subscription will be present.
    - On this branch, all your subscriptions will be listed.

Between testing on `trunk` and this branch, it's best to wipe your site of subscriptions and to delete the cache (see sql query below), so that the testing is consistant (no cache -> cache of 1 -> cache of 2).

| Branch        | Checkouts (2 orders) | My Account → Subscriptions  |
|:------------- |:-----:|:-----:|
| `trunk`     | <img width="1367" alt="Screen Shot 2022-11-25 at 11 49 18 am" src="https://user-images.githubusercontent.com/8490476/203884283-42407d7a-4ca9-4953-bd71-51f93191e93b.png"> |  <img width="1146" alt="Screen Shot 2022-11-25 at 11 51 12 am" src="https://user-images.githubusercontent.com/8490476/203884425-5bb2bac2-b8a0-457e-8aa4-537c802da5b7.png"> 🔴  only 1 subscription listed |
| This branch  |  <img width="1367" alt="Screen Shot 2022-11-25 at 11 49 18 am" src="https://user-images.githubusercontent.com/8490476/203884283-42407d7a-4ca9-4953-bd71-51f93191e93b.png">  | <img width="1103" alt="Screen Shot 2022-11-25 at 11 56 10 am" src="https://user-images.githubusercontent.com/8490476/203884898-e907a2b5-340e-4bc1-805c-f1244562e820.png"> both subscriptions listed 👍 |


## Product impact
